src/core/algorithms/gfd/balancer.cpp

```diff
@@ -0,0 +1,157 @@
+#include "balancer.h"
+
+#include <algorithm>
+#include <map>
+#include <numeric>
+#include <vector>
+
+std::vector<std::vector<int>> Balancer::balance(const std::vector<int>& weights,
+                                                const int& processors_num) {
+    std::size_t m = std::min(processors_num, (int)weights.size());
+    std::vector<std::vector<int>> result = {};
+    if (weights.begin() == weights.end()) {
+        for (int i = 0; i < processors_num; ++i) {
+            std::vector<int> temp = {};
+            result.push_back(temp);
+        }
+        return result;
+    }
+    for (std::size_t i = 0; i < m; ++i) {
+        // the first value is index
+        std::vector<int> temp = {static_cast<int>(i)};
+        result.push_back(temp);
+    }
+    // fill processors initially
+    // count optimal
+    double optimal = 0;
+    std::size_t i = 0;
+    for (const int& weight : weights) {
+        result.at(i++).push_back(weight);
+        i = i == m ? 0 : i;
+        optimal += weight;
+    }
+    optimal /= m;
+    // sort processors (for convenience)
+    for (std::vector<int>& processor : result) {
+        std::sort(processor.begin() + 1, processor.end());
+    }
+    // ALGORITHM
+    // 1st step

```

Comment:
Try to break down this huge method in a bunch of smaller ones. Each code block denoted by a comment looks like a good candidate for extraction to a separate method.